### ISSUE TITLE
Refining which section of a ship fired upon is being hit

### ISSUE THREAD

**Issue by AncientSion** [Aatu/FieryVoid#135]
Marcin did add a few functions to narrow down which section of a ship is being targeted.
Currently, the algorithm will always pick a "narrow" arc over a wide arc, i.e. when, by B5Wars rules,  the DEFENDER could decide wether to hit Front or Bord/Starboard, FV will make the choice for you, it will always pick FRONT section, or AFT section, etc. This makes ships more fragile when compared to B5Wars.

Marcin suggested to refine the search in a way that will compare all available sections to each other, and then it will pick the more "resistent" structure section to apply the shots to, making shups more enduring than they are now, occasionally by far.

The general idea, and that is very good, quick way of describing it:,

When trying where to hit a ship, the game should, to the advantage of the defender, always try to make a ship as tough as possible, i.e.  expose "intact" sections to fire instead of letting damage go through the destroyed FRONT into PRIMARY, when PORT / BOARD would be intact.

When deciding in which order to use weapons fire, the game should, do the advantage of the attacker, try to maximize damage, i.e. apply weapons in the most efficient, dmg-wise, order.

So, marcin suggested this:
```
public doFillSectionListWithStruct($sectionList){
               foreach($sectionList as $key => $currSection) {
                              if($currSection["boxes"] == 0){ //else does not need filling
				   $structure = $this->getStructureSystem($location);
                		   if ($structure != null && $structure->isDestroyed($turn-1))
                                            $sectionList[$key]["boxes"] = ?; //section remaining Structure
                                            $sectionList[$key]["armor"] = ?; //section Structure Armor
				   }
                              }
               }
}
 
 
public doGetSectionList(){ //has to be redefined for each class with different arcs
               //VIRTUAL outer sections: ID, arc from, arc to, profile, current structure boxes, current armor
               $sectionList = array(
                              array("id"=>1, "arcFrom"=>330, "arcTo"=>30, "profile"=>$this->$forwardDefense, "boxes"=>0, "armor"=>0),
                              array("id"=>2, "arcFrom"=>150, "arcTo"=>210, "profile"=>$this->$forwardDefense, "boxes"=>0, "armor"=>0),
                              array("id"=>3, "arcFrom"=>210, "arcTo"=>330,  "profile"=>$this->$sideDefense, "boxes"=>0, "armor"=>0),
                              array("id"=>4, "arcFrom"=>210, "arcTo"=>330,  "profile"=>$this->$sideDefense, "boxes"=>0, "armor"=>0),
               );
               $sectionList = doFillSectionListWithStruct($sectionList);
               return $sectionList;
}//doGetSectionList
 
 
 
public function doGetHitSectionAsArray($relativeBearing){ //decides which section/profile should be hit
               $sectionList = doGetSectionList();
 
            $currLocation = array("id"=>0, "arcFrom"=>0, "arcTo"=>0, "profile"=>99, "boxes"=>0, "armor"=>0);//obviously worse than any real location, even destroyed
		foreach($sectionList as $key => $currSection){
			if( mathlib::isInArc($relativeBearing,$currSection["arcFrom"],$currSection["arcTo"])){ //else not in arc
/*algorithm deciding if $currSection is better target than $currLocation*/
				$isBetter = false;
				$weightedBoxesLocation = $currLocation["boxes"]*(10+$currLocation["armor"]);
				$weightedBoxesSection = $currSection["boxes"]*(10+$currSection["armor"]);
				//if $currSection is >0 structure and $currLocation is 0
				if( ( $weightedBoxesSection > 0 ) and ($weightedBoxesLocation == 0) ) $isBetter = true;
				//if profile is no worse and weighted structure is bigger
				if( ( $weightedBoxesSection > $weightedBoxesLocation ) 
					and ( $currSection["profile"] <= $currLocation["profile"] )
				) $isBetter = true;
				//if there's at least twice Structure and at least 20 boxes of it, even if profile is bigger
				//else better profile wins, unless destroyed and current is not
				if( ( $weightedBoxesSection >= (2*$weightedBoxesLocation) ) 
					and ( $currSection["boxes"] >= 20 )
				) {$isBetter = true;} else if(
					($currSection["profile"] < $currLocation["profile"])
					and ( ( $currSection["boxes"] > 0 ) or ( $currLocation["boxes"] = 0 ) )
				) {$isBetter = true;}
				

				if($isBetter==true){
					$currLocation=$currSection;
				}
			}
		}
                
            return $currLocation;
} //doGetHitSectionAsArray


public function getRelativeBearing($pos, $shooter, $turn, $weapon){

            $tf = $this->getFacingAngle();
            $shooterCompassHeading = 0;
            
            if ( ($weapon == null) || (! $weapon->ballistic)){
                $shooterCompassHeading = mathlib::getCompassHeadingOfShip($this, $shooter);
            }else{
                $shooterCompassHeading = mathlib::getCompassHeadingOfPos($this, $pos);
            }

	$relativeBearing = Mathlib::addToDirection($tf,$shooterCompassHeading);
	if( Movement::isRolled($this) ){ //if ship is rolled, mirror relative bearing
		if( $relativeBearing <> 0 ) { //mirror of 0 is 0
			$relativeBearing = 360-$relativeBearing;
		}
	}

	return $relativeBearing;

} //getRelativeBearing



public function getHitSection($pos, $shooter, $turn, $weapon){
	$relativeBearing = getRelativeBearing($pos, $shooter, $turn, $weapon);

	$hitSectionArray = doGetHitSectionAsArray($relativeBearing);
	$location = $hitSectionArray["id"];

//the rest without change
	if ($location != 0){
                if (!isset($this->hitChart[0])){
                    debug::log("!isset ship->hitchart[0] getHitSection");
                    if ((($this instanceof MediumShip && Dice::d(20)>17 ) || Dice::d(10)>9) && !$weapon->flashDamage){
                        $location = 0;
                    }
                }
              
                $structure = $this->getStructureSystem($location);
                if ($structure != null && $structure->isDestroyed($turn-1))
                    return 0;
        }
        
        return $location;
            
 } //getHitSection



public function doGetDefenceValue($tf, $shooterCompassHeading){
	$relativeBearing = Mathlib::addToDirection($tf,$shooterCompassHeading);
	if( Movement::isRolled($this) ){ //if ship is rolled, mirror relative bearing
		if( $relativeBearing <> 0 ) { //mirror of 0 is 0
			$relativeBearing = 360-$relativeBearing;
		}
	}

	$hitSectionArray = doGetHitSectionAsArray($relativeBearing);
	$profile = $hitSectionArray["profile"]; 
}
```

How does it work ?
In general, the game will check which sections of a ship are elligible for being hit. If its more than one, it compared the sections according to profile, structure armour and structure HP.
According to some weighting, it will then decide which section can endure more damage (incl. profile, i.e. %) and then return that section to apply damage to.

I like it in general, not sure about the whole weighting thing, shouldnt it include the entire sections remaining health instead of only structural health i.E. ?
One thing i dont understand is:

What is the relativeBearing and why we need it ? :D
Secondly, since is done for EACH shot, its quite a bit of memory i guess and we are already reaching the upper treshhold of Aatus virtual server.
On the other hand, the instances where there are actually more than 1 elligible section to be hit are rather rare, so in general it wont need to process and compare too much.

**Comment by AncientSion** [comment]
Also why the **** cant i assign more than one person to an issue.

**Comment by tksolway** [comment]
I think it only allows you to assign one person.  But I like the idea

**Comment by tksolway** [comment]
I think you could simplify the algorithm by simply looking at the remaining structure on the side and ignore armour values, weighting and systems etc.  In most cases the remaining structure will be a good indication of how badly a side has been hit.  Even in cases where that side has just taken a mass cannon hit and lost half its structure with the rest of the side intact.  I'm probably going to take the hit on a side with more structure to negate the chance of another lucky roll stripping the side off before any weapons are actually hit.  

Doing it this way could make the coding for this very easy.  As you would simply have the gethitsection function do a quick check on the percentage of damage to the structure on the two sides it needs to choose from.  I haven't looked at the current code for this, but I assume when the mathlib hits an intersection it just defaults to one side, instead put a quick subfunction in that grabs the structure object of each section and checks the relative percentages, if both are undamaged chose the larger arc (I always choose the side of a ship rather than front or rear in game because of how most ships are laid out all things being equal.)  

What do you think?

**Comment by Zero1627** [comment]
Tom, 
I have written a big mail with answer to this and sent it this morning, but sadly did not save it in 'sent' folder :( . Anyway, I'll try to recap main points here:
 - simplify by skipping systems: consider it done, I never considering going this deep :)
 - simplify by skipping armor: possible but it wouldn't simplify much (delete a grand total of 2 lines), and I think armor is important information.

>I haven't looked at the current code for this, but I assume when the mathlib hits an intersection it just defaults to one side
getHitSection does just that. 

>instead put a quick subfunction in that grabs the structure object of each section (...)
I do just that, just use different algorithm than You propose.

Next post - algorithm details. These are of course open to discussion, consider my algorithm an initial sketch.

**Comment by Zero1627** [comment]
Description of my algorithm - to be considered a base for discussion:

to compare durability of sections, I use section (remaining) structure modified by armor:
armorstructure = structure*(10+armor) . 

disclaimer: there are situations which make any generalizations false. But this is mechanical algorithm - it can't take into account everything. I want something that will work all right in most cases.

Algorithm steps:

1) never choose destroyed section if undestroyed one is available. 
reason: PRIMARY systems (...and Structure too) are mighty important

2) if difference between between durability of sections is small, choose smaller profile
reason: if there's no serious difference in ability to soak damage, choose option which gives greater chance of not being hit in the first place :)
here is of cource question what difference is large enough. I coded it in as 'at least twice as big (armorstructure) and at least 20 structure boxes remaing' (that last one - because if remaining structure is very small, a few boxes do not matter much, even it that's a high percentage)

3) if not already selected, go for larger armorstructure

4) if equal, choose at random (here it's really not random but in order of entering sections)


Tom's proposed changes, as I understand them:
a) use structure instead of armorstructure
b) choose larger arc if in doubt

my opinion:
a) it would simplify little (grand total of 2 lines), with effect of not taking armor into account (which is important, especially for light guns). My algorithm already gives low weight to armor (between armor 4 and 6 there's 14% difference, for equal structure), meaning that if there is signinficant difference in remaining structure - larger structure will be chosen. But for small differences higher armor will be preferred (as secondary choice, if there's no difference in profile).
b) arc defines nothing relevant. Argument about typical layout - well, it's true for some ships, untrue for others. G'Quan, I'd take shooting to side. Omega, to Fwd. Primus, Fwd unless large volume of fire is incoming. All of them I consider pretty typical ships...
I use profile instead: all other things being broadly similar: I prefer not to get hit (or at least have bigger chance).

Consider discussion open :)

**Comment by AncientSion** [comment]
What i want to know is what is up with relativeBearing

**Comment by Zero1627** [comment]
>What i want to know is what is up with relativeBearing

this, too, I answered by email a few days ago :( - must have been lost somewhere...

Relative bearing is angle from target to firing ship, relative to current ship position. 
Current code uses true bearing (relative to map), and in all operations compensates for current ship position, and THEN compensates for roll.
My code derives relative bearing (including rolled status) right in the beginning, and all further computing uses this data, no need for compensations. Computing itself is much easier this way.

**Comment by AncientSion** [comment]
Ok, i see. 

Thanks for elaborating,

**Comment by Zero1627** [comment]
note: right in the beginning of my code there is function doFillSectionListWithStruct. As I did not know proper functions, I left here room for filling 2 variables necessary for my algorithm (remaining structure points and armor); structure object itself I already have, it's $structure variable:

   $sectionList[$key]["boxes"] = ?; //section remaining Structure
   $sectionList[$key]["armor"] = ?; //section Structure Armor

This is (hopefully) the only place in the code that needs to be completed by You (except bugfixing of course).

**Comment by tksolway** [comment]
Sorry Zero, somehow I read in the code that you were adding the structure of all the systems in the calculation as well.  As you explain above I don't see any issue.  As for the last tiebreaker, side or profile, it really just comes down to semantics, and I have to agree that profile at least has a better logical argument behind it.  So to make a long story short, I'm easy. :-)

**Comment by Zero1627** [comment]
thx Tom :)

**Comment by AncientSion** [comment]
This wont work properly at this point due to a fatal flaw in the function chain.
The function chain for resolving fire basic basicly goes like

$weapon->fire @ ship
$targetShip->getHitSystem
$system->doDamage(ToSelf)

With this code, the chain is being modified "in the middle", i.e. @targetShip->getHitSystem
At that point in the chain, a ship has already been hit and we are only determining which system to hit, and hence, which location to hit.
@Zero1627 s functions weight sections versus each other with profile, armour etc.

The flaw is this:
What we are doing, is refining which section the game (scope ship) will choose a system from.
However, before that, when actually determining IF A WEAPON HITS A SHIP, the game will pick the old doGetDefenceValue function to calculate the odds.
So, while it would, after being hit, choose the "toughest" section to hit, it will still always pick FRONT over SIDE when rolling fire orders (for almost all vessels).


```
 public function calculateHit($gamedata, $fireOrder){
        if ($this->ballistic){
            $movement = $shooter->getLastTurnMovement($fireOrder->turn);
            $pos = mathlib::hexCoToPixel($movement->x, $movement->y);
            $defence = $target->getDefenceValuePos($pos);
        }else{
            $defence = $target->getDefenceValue($shooter);
        }

```

```
        public function getDefenceValuePos($pos){
            $tf = $this->getFacingAngle();
            $shooterCompassHeading = mathlib::getCompassHeadingOfPos($this, $pos);
          
            return $this->doGetDefenceValue($tf,  $shooterCompassHeading);
        }
        
        public function getDefenceValue($shooter){
            $tf = $this->getFacingAngle();
            $shooterCompassHeading = mathlib::getCompassHeadingOfShip($this, $shooter);
          
            return $this->doGetDefenceValue($tf,  $shooterCompassHeading);
            
        }
        
        
        public function doGetDefenceValue($tf, $shooterCompassHeading){
            if (mathlib::isInArc($shooterCompassHeading, Mathlib::addToDirection(330,$tf), Mathlib::addToDirection(30,$tf) )){
               return $this->forwardDefense;
            }else if (mathlib::isInArc($shooterCompassHeading, Mathlib::addToDirection(150,$tf), Mathlib::addToDirection(210,$tf) )){
                return $this->forwardDefense;
            }else if (mathlib::isInArc($shooterCompassHeading, Mathlib::addToDirection(210,$tf), Mathlib::addToDirection(330,$tf) )){
                return $this->sideDefense;
            }  else if (mathlib::isInArc($shooterCompassHeading, Mathlib::addToDirection(30,$tf), Mathlib::addToDirection(150,$tf) )){
                return $this->sideDefense;
            } 
                
            return $this->sideDefense;
        }
        

```

**Comment by Zero1627** [comment]
Actually, look at the last function in my code - it's redefinition of doGetDefenceValue ;) . I understand that this is separate call, but my algorithm is deterministic - so it should return profile appropriate to section that would be threatened if this shot hits.

**Comment by AncientSion** [comment]
I see it.
I feel its redundant tho, isnt it ?

Going this way, for each and every shot, we do the whole process of fetching the toughest section TWICE, isnt that true ?

How can we make this less redundant. The whole process of fetching all locations via mathlib comparision, weighing etc is done once for calcing toHit, then again for deciding which system to actually hit, if the ship was hit.
Im certainly no pro, but this does seem very inefficient to me and ideally we / i would want a really elegant solution.
Dont you agree, @Zero1627

Also, in case i didnt mention it yet, im impressed how you were able to get this stuff typed out without recent PHP knowledge. Apart from a few minor syntax errors, it did work (on my server).

**Comment by Zero1627** [comment]
>How can we make this less redundant. 

well, actually  quite easily. 
function doGetDefenceValue() - before return, save section data in appropriate (new) class variable for ship:
>$this->chosenSection = $hitSectionArray["id"];

then, in function getHitSection(), replace first 3 lines:
>    $relativeBearing = getRelativeBearing($pos, $shooter, $turn, $weapon);
    $hitSectionArray = doGetHitSectionAsArray($relativeBearing);
    $location = $hitSectionArray["id"];

with new code:
> $location = $this->chosenSection; 

this would mean that section is chosen when the shot is coming, and later procedure is NOT replied (there's only a check if the chosen section is still there).

**Comment by Zero1627** [comment]
>Also, in case i didnt mention it yet, im impressed how you were able to get this stuff typed out without recent PHP knowledge. Apart from a few minor syntax errors, it did work (on my server).

Thank You, Chris :)

**Comment by AncientSion** [comment]
That does seem pretty good.
We do use calculateHit() to decide which section by location ID to be used for THAT shot.
The shot will hot, or miss.
If if hits, it will immediatly try to further resolve the shot, i.e. getHitSystem and DoDamage.
So by using calculateHit() to enter a variable to the ship object, we can then use that variable for getHitSystem().

When a second shot is fired, we can overwrite the formerly decided location on the ship class.

This is very good.

**Comment by Zero1627** [comment]
I don't know exactly how getHitSystem() works, but certainly it's usable by getHitSection() ;)

**Comment by AncientSion** [comment]
It will work.

**Comment by AncientSion** [comment]
Marcin, the functions above are for capital ships and they would also work for bases.

However, im encountered troubles when it comes to anything with shares sections/profiles.

i.e. for HCV, we would have this

```
public doGetSectionList(){ //has to be redefined for each class with different arcs`

               $sectionList = array(
                              array("id"=>1, "arcFrom"=>270, "arcTo"=>90"profile"=>$this->$forwardDefense, "boxes"=>0, "armor"=>0),
                              array("id"=>2, "arcFrom"=>90, "arcTo"=>270, "profile"=>$this->$forwardDefense, "boxes"=>0, "armor"=>0),
               );
               $sectionList = doFillSectionListWithStruct($sectionList);
               return $sectionList;
}
````

As you can see, we have a problem here. Because we define the profile with a static value, while depending on angle, if within 270 and 90, the profile would be either SIDE or FRONT.
This would mean for any type/subtype of HCVs, we would need to further refine the arrays and pending on the bearing, apply forward or side which would greatly increase the code in length (and hence, manual typign).


Did you consider this any where because i dont think you did (nether did i) ?

**Comment by AncientSion** [comment]
```
            $locs[] = array("loc" => 1, "min" => 270, "max" => 330, "profile" => $this->sideDefense);
            $locs[] = array("loc" => 1, "min" => 330, "max" => 30, "profile" => $this->forwardDefense);
            $locs[] = array("loc" => 1, "min" => 30, "max" => 90, "profile" => $this->sideDefense);

            $locs[] = array("loc" => 2, "min" => 90, "max" => 150, "profile" => $this->sideDefense);
            $locs[] = array("loc" => 2, "min" => 150, "max" => 210, "profile" => $this->forwardDefense);
            $locs[] = array("loc" => 2, "min" => 210, "max" => 270, "profile" => $this->sideDefense);
```

i guess this would work, so instead of 2 profiles, we build up to 6 profiles for HCVs ?

**Comment by AncientSion** [comment]
'''
            $locs[] = array("loc" => 4, "min" => 0, "max" => 30, "profile" => $this->forwardDefense);
            $locs[] = array("loc" => 4, "min" => 30, "max" => 150, "profile" => $this->sideDefense);
            $locs[] = array("loc" => 4, "min" => 150, "max" => 180, "profile" => $this->forwardDefense);

            $locs[] = array("loc" => 3, "min" => 150, "max" => 180, "profile" => $this->forwardDefense);
            $locs[] = array("loc" => 3, "min" => 180, "max" => 210, "profile" => $this->sideDefense);
            $locs[] = array("loc" => 3, "min" => 210, "max" => 360, "profile" => $this->forwardDefense);
'''

proper redefinition to aft-less drazi cap ?

**Comment by AncientSion** [comment]
```

            $locs[] = array("loc" => 1, "min" => 330, "max" => 30, "profile" => $this->forwardDefense);
            $locs[] = array("loc" => 1, "min" => 30, "max" => 150, "profile" => $this->sideDefense);
            $locs[] = array("loc" => 1, "min" => 150, "max" => 210, "profile" => $this->sideDefense);
            $locs[] = array("loc" => 1, "min" => 210, "max" => 360, "profile" => $this->forwardDefense);
```

stock MCV

**Comment by AncientSion** [comment]
```
            $locs[] = array("loc" => 4, "min" => 0, "max" => 30, "profile" => $this->forwardDefense);
            $locs[] = array("loc" => 4, "min" => 30, "max" => 150, "profile" => $this->sideDefense);
            $locs[] = array("loc" => 4, "min" => 150, "max" => 180, "profile" => $this->forwardDefense);
            $locs[] = array("loc" => 3, "min" => 180, "max" => 210, "profile" => $this->forwardDefense);
            $locs[] = array("loc" => 3, "min" => 210, "max" => 330, "profile" => $this->sideDefense);
            $locs[] = array("loc" => 3, "min" => 330, "max" => 360, "profile" => $this->forwardDefense);
```

drazi medium ?

**Comment by Zero1627** [comment]
>As you can see, we have a problem here. Because we define the profile with a static value, while depending on angle, if within 270 and 90, the profile would be either SIDE or FRONT.
This would mean for any type/subtype of HCVs, we would need to further refine the arrays and pending on the bearing, apply forward or side which would greatly increase the code in length (and hence, manual typign).
Did you consider this any where because i dont think you did (nether did i) ?

I did consider that and did comment in mail I sent You. Every separate section/profile pair should be considered a separate entity for the algorithm. So a typical HCV (or MCV, for that matter) would have 6 'virtual sections' which should be defined in doGetSectionList() - I see You separately invented the same idea :)


in fact, I even left a comment in the function itself - the first line is:
//VIRTUAL outer sections: ID, arc from, arc to, profile, current structure boxes, current armor

standard HCV:
$sectionList = array(
   array("id"=>1, "arcFrom"=>330, "arcTo"=>30, "profile"=>$this->$forwardDefense, "boxes"=>0, "armor"=>0),
   array("id"=>1, "arcFrom"=>30, "arcTo"=>90, "profile"=>$this->$sideDefense, "boxes"=>0, "armor"=>0),
   array("id"=>1, "arcFrom"=>270, "arcTo"=>330, "profile"=>$this->$sideDefense, "boxes"=>0, "armor"=>0),
   array("id"=>2, "arcFrom"=>90, "arcTo"=>150, "profile"=>$this->$sideDefense, "boxes"=>0, "armor"=>0),
   array("id"=>2, "arcFrom"=>150, "arcTo"=>210, "profile"=>$this->$forwardDefense, "boxes"=>0, "armor"=>0),
   array("id"=>2, "arcFrom"=>210, "arcTo"=>270, "profile"=>$this->$sideDefense, "boxes"=>0, "armor"=>0),
);

Drazi-style HCV:
$sectionList = array(
   array("id"=>3, "arcFrom"=>180, "arcTo"=>210, "profile"=>$this->$forwardDefense, "boxes"=>0, "armor"=>0),
   array("id"=>3, "arcFrom"=>210, "arcTo"=>330, "profile"=>$this->$sideDefense, "boxes"=>0, "armor"=>0),
   array("id"=>3, "arcFrom"=>330, "arcTo"=>360, "profile"=>$this->$forwardDefense, "boxes"=>0, "armor"=>0),
   array("id"=>4, "arcFrom"=>0, "arcTo"=>30, "profile"=>$this->$forwardDefense, "boxes"=>0, "armor"=>0),
   array("id"=>4, "arcFrom"=>30, "arcTo"=>150, "profile"=>$this->$sideDefense, "boxes"=>0, "armor"=>0),
   array("id"=>4, "arcFrom"=>150, "arcTo"=>180, "profile"=>$this->$forwardDefense, "boxes"=>0, "armor"=>0),
);

For MCVs, arc should be the same as HCVs, but "boxes" and "armor" should be "1" , so they're not filled automatically by doFillSectionListWithStruct()  (no point, all of them would have the same remaining structure and armor ;) ) - NOT like You wrote above, because MCVs technically do have outer sections (outer hit tables) and they should be chosen instead of PRIMARY hit table. Your table labeled 'stock MCV' should be proper for LCV, though.


And finally, Drazi-style capital ship (without Aft) would look like this:
$sectionList = array(
   array("id"=>1, "arcFrom"=>330, "arcTo"=>30, "profile"=>$this->$forwardDefense, "boxes"=>0, "armor"=>0),
   array("id"=>3, "arcFrom"=>210, "arcTo"=>330,  "profile"=>$this->$sideDefense, "boxes"=>0, "armor"=>0),
   array("id"=>3, "arcFrom"=>180, "arcTo"=>210,  "profile"=>$this->$forwardDefense, "boxes"=>0, "armor"=>0),
   array("id"=>4, "arcFrom"=>30, "arcTo"=>150,  "profile"=>$this->$sideDefense, "boxes"=>0, "armor"=>0),
   array("id"=>4, "arcFrom"=>150, "arcTo"=>180,  "profile"=>$this->$forwardDefense, "boxes"=>0, "armor"=>0),

);



I believe these are all necessary redefinitions except bases?

**Comment by AncientSion** [comment]
Thanks.

I DID read your mail and in fact i even read the comment / whole code earlier today when i stumbled upon that particular problem.
The comment really didnt make much sense to me tho, Virtual Outer Sections combined with the 4 obvious location objects didnt make any sense, sorry.

At this point, i still dont understand why you would think my MCV setup would be wrong.

MCVs only have primary structure and technically they have 4 profiles, the same value each at topmost and bottommost 60 degree, the rest at the sides 2*60 degree.
So i guess a setup with 4 location objects is correct, the only difference being in arc and profile, so the game can correctly calcuate the odds for hitting, because everything else is the same regardless of where you fire from, is it not ?
If you want to tell me that for MCVs, instead of comparing front/aft "structure", which doesnt exist,  we need to compare all systems remaining Health (instead of Structure), then you need to be more clear :D
If that is what you are saying, a sub function for MCVs need to bedeclared (where, instead of only checking structure hp/armour, the game would instead sum up systems hp/medium armor.

```
 $locs[] = array("loc" => 0, "min" => 330, "max" => 30, "profile" => $this->forwardDefense);
            $locs[] = array("loc" => 0, "min" => 30, "max" => 150, "profile" => $this->sideDefense);
            $locs[] = array("loc" => 0, "min" => 150, "max" => 210, "profile" => $this->sideDefense);
            $locs[] = array("loc" => 0, "min" => 210, "max" => 360, "profile" => $this->forwardDefense);
```

**Comment by tksolway** [comment]
One thing i think we forgot. Once weve made a choice about which side to take the fire on. It cannot change for fire from that target. This algorithm will have to check the state at the beginning of the turn not state after accepting dmg. Othwise it will end up spreading the damage over two sections in a turn.

Thanks,
Thomas Solway B.Sc.

On Jan 9, 2016, at 5:05 PM, AncientSion <notifications@github.com<mailto:notifications@github.com>> wrote:


Thanks.

I DID read your mail and in fact i even read the comment / whole code earlier today when i stumbled upon that particular problem.
The comment really didnt make much sense to me tho, Virtual Outer Sections combined with the 4 obvious location objects didnt make any sense, sorry.

At this point, i still dont understand why you would think my MCV setup would be wrong.

MCVs only have primary structure and technically they have 4 profiles, the same value each at topmost and bottommost 60 degree, the rest at the sides 2*60 degree.
So i guess a setup with 4 location objects is correct, the only difference being in arc and profile, so the game can correctly calcuate the odds for hitting, because everything else is the same regardless of where you fire from, is it not ?
If you want to tell me that for MCVs, instead of comparing front/aft "structure", which doesnt exist, we need to compare all systems remaining Health (instead of Structure), then you need to be more clear :D
If that is what you are saying, a sub function for MCVs need to bedeclared (where, instead of only checking structure hp/armour, the game would instead sum up systems hp/medium armor.

 $locs[] = array("loc" => 0, "min" => 330, "max" => 30, "profile" => $this->forwardDefense);
            $locs[] = array("loc" => 0, "min" => 30, "max" => 150, "profile" => $this->sideDefense);
            $locs[] = array("loc" => 0, "min" => 150, "max" => 210, "profile" => $this->sideDefense);
            $locs[] = array("loc" => 0, "min" => 210, "max" => 360, "profile" => $this->forwardDefense);


-

**Comment by AncientSion** [comment]
I think we do it like this on purpose, i.e. the sole reason for this thing is so that the ship will expose the strongest section HIT AFTER HIT on purpose.
At least thats what i think is the background for Marcin's initial push :D ?

And yes, currently it will definitly spread over up to 3 sections (base) within a single turn, thats current state and intended (as far as im concerned).

@Zero1627  ?

**Comment by Zero1627** [comment]
Wow, so many issues to comment on! 

>I believe these are all necessary redefinitions except bases?
I was mistaken here - I skipped fighters. I believe they should be like Caps, but with nominal Structure and Armor instead of dynamic.

>The comment really didnt make much sense to me tho, Virtual Outer Sections combined with the 4 obvious location objects tied to Cap didnt make any sense at that point.
Sorry, I wasn't precise enough.

>MCVs only have primary structure
Yes, but they do have outer hit tables, and the result should lead there - definitely not straight to PRIMARY table :) . Hence the list should be exactly the same as for HCVs, but no need of getting actual structure and armor values (they'll be the same as there's only one structure).

>If you want to tell me that for MCVs, instead of comparing front/aft "structure", which doesnt exist, we need to compare all systems remaining Health
No I don't. But I do want to tell You that it matters whether we use Front or Aft or Primary hit table.

**Comment by Zero1627** [comment]
>Once weve made a choice about which side to take the fire on. It cannot change for fire from that target. This algorithm will have to check the state at the beginning of the turn not state after accepting dmg. Othwise it will end up spreading the damage over two sections in a turn.

I haven't missed that fact, and I hgave thought about it, and have arguments for my aproach that in my mind are serious.



First, let's split it into 2 separate issues:

>ISSUE 1: player has to decide about allocation of fire before any rolling, my algorithm does that on the fly, before each shot.

Reasoning: player does quite complex analysis: 
 - state of target ship, importance of systems in current situation ('I need those Lasers that are about to recharge, damn the side!'; 'Lasers won't recharge for ages anyway, without TAs fighters will eat me - damn the Fwd!')
 - incoming weapons ('I have only 1 box of structure remaining - so what, armor 6 will stop those Scatter Pulsars just fine...')
 - predicted outcome ('yes, side is more durable... but that Mag Gun will kill it anyway. Take it on nonexisting Fwd, maybe it'll miss...')
Algorithm's analysis is much more rudimentary (takes into account just own profile and structure and armor, and that's it). Algorithm that would try to predict result beforehand would have to be either very complex (to take many things into account), or extremely simple (one here, one there...).

So: Player makes educated decision on estimated data. Algorithm makes simple decision on actual data. I expect the result to be worse that players' decision anyway - so in my mind it's acceptable approximation, even if based on technically illegal data.



>ISSUE 2: player allocates fire from entire unit, my algorithm allocates fire for each weapon separately.  

Reasoning: to sensibly allocate fire from entire unit, we're back at complex approach that will estimate incoming fire and its results. 
As I do not see us creating such algorithm, We're back to assigning for each shot. 

I think it's acceptable because:
1) decisions are worse than player's anyway
2) on average case, net result would be similar to more-or-less-sensibly allocating fire of entire units. 
EXCEPTION: fire from one unit would be split when it should not be. 

I accept this exception on the basis that in most cases result is much ckloser to what it should be. If You don't, algorithm should be changed to checking just profile and choosing smaller (or if equal, alternating between ships). Result would be much worse in this case, but still better than nothing, and not creating the problematic exception.

**Comment by AncientSion** [comment]
I like how the ships and bares are way more endurable now when earlier you would just fly straight into that one hex to fire into the open primary. I think its a good change for the gameplay.
However, i have no opinion if this is actually a change that moves FV away from B5Wars or closer to it. i WAS under the impression that it would move us closer to B5Wars when i began the task, but i really dont care either way anymore :p

**Comment by tksolway** [comment]
Checking for damage levels at each individual fire step won't work.  When choosing a hit location on a Spine hit, the target must take all the fire from a single source on the same side.  Splitting the fire like this would make ships "tougher".  For example we have a large ship, say a Sharlin or an Omega that manages to bring most of its weapons to bear down the spine of some poor unsuspecting HCV.  In this situation I've got a good chance to kill this ship outright.  If it has taken no damage previously, then I have to at a minimum get through two structure blocks either the forward or aft (up to player or engine) and then the primary.  However, if we go ahead with this change, suddenly I will need to get through three structure blocks, since after each weapon is fired the damage gets switched from forward to aft and then back again.  It makes sense that the player or engine will choose the section best able to take the hit, but the rule specifically prevents him from spreading that damage out for this reason.  We have to make sure any algorithm we implement does this properly, or we will be introducing a semi-secret trick into the game that will be a serious advantage to anyone that figures out how to use it.

**Comment by Zero1627** [comment]
In this case, I suggest reducing weighing part:
>//if $currSection is >0 structure and $currLocation is 0
                if( ( $weightedBoxesSection > 0 ) and ($weightedBoxesLocation == 0) ) $isBetter = true;
                //if profile is no worse and weighted structure is bigger
                if( ( $weightedBoxesSection > $weightedBoxesLocation ) 
                    and ( $currSection["profile"] <= $currLocation["profile"] )
                ) $isBetter = true;
                //if there's at least twice Structure and at least 20 boxes of it, even if profile is bigger
                //else better profile wins, unless destroyed and current is not
                if( ( $weightedBoxesSection >= (2*$weightedBoxesLocation) ) 
                    and ( $currSection["boxes"] >= 20 )
                ) {$isBetter = true;} else if(
                    ($currSection["profile"] < $currLocation["profile"])
                    and ( ( $currSection["boxes"] > 0 ) or ( $currLocation["boxes"] = 0 ) )
                ) {$isBetter = true;}

To just checking profile:
>if($currSection["profile"] < $currLocation["profile"]) {$isBetter = true;}

This way it will NOT split any fire (be fully deterministic), but at least will choose lesser profile (instead of current always front profile. It would still be improvement over current state.

**Comment by tksolway** [comment]
Perhaps there could be something that checks if that section was destroyed on a previous turn as well.  It should never choose to take the fire on a destroyed section when a non-destroyed section exists.

**Comment by Zero1627** [comment]
yes, such test may be added.

**Comment by tksolway** [comment]
solved in #144
